### DISCUSSION TITLE
Building a kernel extension for macOS using bazel

### DISCUSSION THREAD

**Discussion by YungRaj** [bazelbuild/rules_apple#2562] [Q&A] [answered]
Hey all,

 Is there a way to build kernel extensions for macOS using bazel?

I want to port https://github.com/YungRaj/MacRootKit to bazel so that I don't have to deal with the stress that comes with using another build system such as cmake.  Any ideas?

I know that `genrule()`'s could probably work in building the `.kext` artifact but I think the major challenge is getting the Xcode frameworks and tooling to play nicely and ensuring that all the relevant compiler and linker flags (which I use here https://github.com/YungRaj/MacRootKit/blob/main/make_kext.mk) are included.

Thanks

**Comment by brentleyjones** [comment]
I haven't used it personally, but the [`macos_kernel_extension`](https://github.com/bazelbuild/rules_apple/blob/master/doc/rules-macos.md#macos_kernel_extension) rule exists.

**Reply by keith** [reply]
oop thanks, I totally missed this an recommended OP file an issue

**Comment by YungRaj** [comment]
Okay, great. Thanks! I'll give a try and report back.

It looks like the first thing I noticed is that I cannot specify the target architecture for the kext that I want to build.

Currently, MacRootKit supports `arm64` and `x86_64`. Is that supported?

And maybe, using a special rule like `macos_kernel_extension()` is not necessarily needed? Would it work if I port all the existing linker and compiler flags that I used via `Makefile` and create the `BUILD` target that way? Does this cause issues with Xcode's tools playing nicely?

Thanks!

**Reply by keith** [reply]
in bazel the architecture is managed via `--macos_cpus`, so you can build a fat version that way by passing `--macos_cpus=x86_64,arm64`.

> And maybe, using a special rule like macos_kernel_extension() is not necessarily needed? Would it work if I port all the existing linker and compiler flags that I used via Makefile and create the BUILD target that way?

If the flags you're referring too only existed to make the extension work at all, the benefit of the use case specific rule is to abstract that so ideally you don't have to provide them yourself just to make it work. The rules also manage how the binary is packaged, which is likely unique to this rule.

**Reply by YungRaj** [reply]
Generally speaking I don't need to worry about code signing my kext, so that was why I was brainstorming a path forward without using the rule. I guess since all the code signing parameters to the `macos_kernel_extension()` are optional it's not a big blocker. Thanks so much for the help! I'll report back when I get something working.

**Reply by keith** [reply]
probably easier to strip the codesigning from the final binary

**Reply by YungRaj** [reply]
I have the kernel extension's code built using bazel. For some reason the `cc_library()` static library that I have as a dependency to `macos_kernel_extension()` doesn't link in the dependency that contains all the kext's code. I should be able to figure this out and once I do, I'll test the final kext artifact on baremetal and in a virtual machine.

**Reply by keith** [reply, marked as answer]
i don't know how kernel extensions work but maybe you need `alwayslink=True` on the `cc_library` to make sure the contents are loaded?

**Reply by YungRaj** [reply]
that worked. I still need to finish up certain rules, but I'm pretty sure the rest of the porting to bazel should work fine. thanks for all the help!
